[TP][DTensor Perf] Some perf improvement to reduce DTensor CPU overhead

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106524

By inspecting a small TP benchmark, we found couple things we can optimize:
1. We call deep_copy so many times when we initialize DTensor.
2. Some shading_prop is not cached successfully.
3. We are still calling redistribute when not necessary.

![image](https://github.com/pytorch/pytorch/assets/6937752/b847d110-eea1-45df-9298-066d0ba07dd7)

![image](https://github.com/pytorch/pytorch/assets/6937752/fc08f564-caed-496b-80d7-275c1dba3806)

![image](https://github.com/pytorch/pytorch/assets/6937752/fdc06cc4-a4ba-48e8-a118-c041bbd04f5e)


So we want to:
1. Remove the deep_copy, and we now make placements a tuple so we are sure it's immutable.
2. Somehow the op_schema gets changed during sharding_op propogation, so we store a hash version of it before passing it to sharding_prop. Ideally we want to figure out why `op_schema` gets changed, but looks like in both index and detach/view op, all get changed, it might take more time to debug.
3. Also when we do hashing of op_schema, we want to hash the entire args_schema not just the args_spec which only contains the DTensorSpec from args which are Dtensors.
4. It turns out that sometimes, DTensor has mem_format to be None (not contiguous) and this will lead to redistribute get triggered, so that we only need to compare type/shape and stride in the metadata.

Also we need to ensure _Partial and Shard have different hash value in the DTensorSpec.

![image](https://github.com/pytorch/pytorch/assets/6937752/321e6890-1ab6-4975-adc9-524c6ef9a76b)

fduwjj committed:
DTensor CPU overhead reduction

[ghstack-poisoned]

fduwjj committed:
Update on "[RFC] Some perf improvement to reduce DTensor CPU overhead"


By inspecting a small TP benchmark, we found couple things we can optimize:
1. We call deep_copy so many times when we initialize DTensor.
2. Some shading_prop is not cached successfully.
3. We are still calling redistribute when not necessary.

![image](https://github.com/pytorch/pytorch/assets/6937752/b847d110-eea1-45df-9298-066d0ba07dd7)

![image](https://github.com/pytorch/pytorch/assets/6937752/fc08f564-caed-496b-80d7-275c1dba3806)

![image](https://github.com/pytorch/pytorch/assets/6937752/fdc06cc4-a4ba-48e8-a118-c041bbd04f5e)


So we want to:
1. Remove the deep_copy, this is something risky so it's RFC here.
2. Somehow the op_schema gets changed during sharding_op propogation, so we store a string version of it before passing it to sharding_prop. Ideally we want to figure out why `op_schema` gets changed, but looks like in both index and detach/view op, all get changed, it might take more time to debug.
3. It turns out that Sometimes, DTensor has mem_format to be None (not contiguous) and this will lead to redistribute get triggered.


[ghstack-poisoned]

XilunWu commented:
skimming through `view_ops.py` and it seems the sharding prop does modify the input `op_schema` in some cases. \[code pointer](https://github.com/pytorch/pytorch/blob/2d4b1ae434d28ed66b3c96cd47e21d648bd6a0b8/torch/distributed/_tensor/ops/view_ops.py#L625-L633)

wanchaol commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -390,7 +390,13 @@ def __eq__(self, __o: object) -> bool:
             and self.placements == __o.placements
-            and self.tensor_meta == __o.tensor_meta
+            # perf hack to avoid redistribute due to memory_format to be None.
+            and self.tensor_meta.shape == __o.tensor_meta.shape
```
probably we should refactor tensor_meta to only take tensor shape/stride, and drop other fields (sharding prop only need shape anyways). This can be done in follow up PRs.

fduwjj replied:
I see. A good point. Do we want to keep the rest in a separate field or we want to completely drop it here? because this metadata is defined in torch.fx. Maybe we only compare the shape here?

wanchaol replied:
In order to do this, I think we can introduce another data structure (maybe call it ShapeSpec and only save shape/stride), then not use the metadata from torch.fx.

If we want to not get rid of the metadata in this PR, we can still only compare the shape here in dtensor spec `__eq__`

wanchaol reviewed (commented):
nice work! see comments inlined, I think we should continue use hash unless `str` give something significantly faster

wanchaol commented:
> cc @wanchaol is there any reason why we have it? Shall we force every sharding prop rule to make `op_schema` immutable?

the view op changes the input args to sharded version when necessary so that the local op are called correctly, so we can't remove that logic. However we might can try making the op schema change as a resharding instead of inplace modify the schema

fduwjj commented on torch/distributed/_tensor/api.py:
```diff
@@ -203,7 +203,7 @@ def __new__(
         # deepcopy and set spec
         r._spec = DTensorSpec(
-            device_mesh, copy.deepcopy(placements), tensor_meta=tensor_meta
+            device_mesh, placements, tensor_meta=tensor_meta
```
I noticed that using `tuple()` also have some cost there.

wanchaol replied:
hmmm how much cost of `tuple` could have? I thought it should be very minimal. another option is to change `from_local`, `distribute_tensor` to call tuple, and we leave `placements` here without any conversion

fduwjj replied:
I like the latter better.

fduwjj commented on torch/distributed/_tensor/sharding_prop.py:
```diff
@@ -276,11 +276,12 @@ def propagate(self, op_overload: OpOverload, op_schema: OpSchema) -> OutputShard
         """
-        if op_schema in self.cached_prop_results:
-            return self.cached_prop_results[op_schema]
+        op_schema_key = str(op_schema)
```
yes agree. Upon checking the trace, str is also expensive. We want to finally get rid of this str but somehow need to ensure `op_schema` does not get changed.

wanchaol replied:
op_schema get changed because of `tensor_meta` got changed? if we change the `__eq__` to only verify tensor_meta shape/stride, would it still be an issue?

fduwjj replied:
No, not `tensor_meta` what I have seen is the placements also get changed.

fduwjj commented:
@wanchaol is there any way that we can make `op_schema` immutable? Looks like this is a tricky thing in python...

fduwjj commented:
@XilunWu not just view op, I saw detach/index op also have the `op_schema` changed.

wanchaol commented:
> @wanchaol is there any way that we can make `op_schema` immutable? Looks like this is a tricky thing in python...

The principal way to make it immutable is to use a frozen data class in python. But we'll need to first change how view op behaves before doing it, as view op currently intentionally change the op schema

fduwjj committed:
Update on "[TP][DTensor Perf] Some perf improvement to reduce DTensor CPU overhead"


By inspecting a small TP benchmark, we found couple things we can optimize:
1. We call deep_copy so many times when we initialize DTensor.
2. Some shading_prop is not cached successfully.
3. We are still calling redistribute when not necessary.

![image](https://github.com/pytorch/pytorch/assets/6937752/b847d110-eea1-45df-9298-066d0ba07dd7)

![image](https://github.com/pytorch/pytorch/assets/6937752/fc08f564-caed-496b-80d7-275c1dba3806)

![image](https://github.com/pytorch/pytorch/assets/6937752/fdc06cc4-a4ba-48e8-a118-c041bbd04f5e)


So we want to:
1. Remove the deep_copy, this is something risky so it's RFC here.
2. Somehow the op_schema gets changed during sharding_op propogation, so we store a string version of it before passing it to sharding_prop. Ideally we want to figure out why `op_schema` gets changed, but looks like in both index and detach/view op, all get changed, it might take more time to debug.
3. It turns out that Sometimes, DTensor has mem_format to be None (not contiguous) and this will lead to redistribute get triggered.


[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/api.py:
```diff
@@ -203,7 +202,7 @@ def __new__(
         # deepcopy and set spec
         r._spec = DTensorSpec(
-            device_mesh, copy.deepcopy(placements), tensor_meta=tensor_meta
+            device_mesh, placements, tensor_meta=tensor_meta  # type: ignore[arg-type]
```
why we have this mypy issue now? should we type-cast to tuple?

wanchaol commented on torch/distributed/_tensor/sharding_prop.py:
```diff
@@ -269,18 +269,19 @@ def __init__(self, propagator: ShardingPropagator) -> None:
         # cache table for sharding propagation results, we might need to
         # limit the size of the cache table in the future
-        self.cached_prop_results: Dict[OpSchema, OutputSharding] = {}
+        self.cached_prop_results: Dict[str, OutputSharding] = {}
```
it looks like we are changing `op_schema` equal and in the meanwhile we are using str to index  and cache the proper results, I feel this two thing might be mutual exclusive. I feel we should still use hash instead of str

fduwjj replied:
No, the op_schema gets changed in propagation. Before we figure out a way to not change it, can we temporarily use str conversion?

fduwjj committed:
Update on "[TP][DTensor Perf] Some perf improvement to reduce DTensor CPU overhead"


By inspecting a small TP benchmark, we found couple things we can optimize:
1. We call deep_copy so many times when we initialize DTensor.
2. Some shading_prop is not cached successfully.
3. We are still calling redistribute when not necessary.

![image](https://github.com/pytorch/pytorch/assets/6937752/b847d110-eea1-45df-9298-066d0ba07dd7)

![image](https://github.com/pytorch/pytorch/assets/6937752/fc08f564-caed-496b-80d7-275c1dba3806)

![image](https://github.com/pytorch/pytorch/assets/6937752/fdc06cc4-a4ba-48e8-a118-c041bbd04f5e)


So we want to:
1. Remove the deep_copy, this is something risky so it's RFC here.
2. Somehow the op_schema gets changed during sharding_op propogation, so we store a string version of it before passing it to sharding_prop. Ideally we want to figure out why `op_schema` gets changed, but looks like in both index and detach/view op, all get changed, it might take more time to debug.
3. It turns out that Sometimes, DTensor has mem_format to be None (not contiguous) and this will lead to redistribute get triggered.


[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -381,16 +381,25 @@ def __hash__(self) -> int:
         # fields to them,
         if self.tensor_meta is not None:
-            return hash((self.mesh, tuple(self.placements), self.tensor_meta.shape))
+            return hash(((self.mesh, self.placements), self.tensor_meta.shape))
```
nit: can hash three together instead of two tuples
```
hash((self.mesh, self.placements, self.tensor_meta.shape))
```

fduwjj replied:
sure, I was about to ask why we do something like this.

fduwjj committed:
Update on "[TP][DTensor Perf] Some perf improvement to reduce DTensor CPU overhead"


By inspecting a small TP benchmark, we found couple things we can optimize:
1. We call deep_copy so many times when we initialize DTensor.
2. Some shading_prop is not cached successfully.
3. We are still calling redistribute when not necessary.

![image](https://github.com/pytorch/pytorch/assets/6937752/b847d110-eea1-45df-9298-066d0ba07dd7)

![image](https://github.com/pytorch/pytorch/assets/6937752/fc08f564-caed-496b-80d7-275c1dba3806)

![image](https://github.com/pytorch/pytorch/assets/6937752/fdc06cc4-a4ba-48e8-a118-c041bbd04f5e)


So we want to:
1. Remove the deep_copy, this is something risky so it's RFC here.
2. Somehow the op_schema gets changed during sharding_op propogation, so we store a string version of it before passing it to sharding_prop. Ideally we want to figure out why `op_schema` gets changed, but looks like in both index and detach/view op, all get changed, it might take more time to debug.
3. It turns out that Sometimes, DTensor has mem_format to be None (not contiguous) and this will lead to redistribute get triggered.


[ghstack-poisoned]

fduwjj committed:
Update on "[TP][DTensor Perf] Some perf improvement to reduce DTensor CPU overhead"


By inspecting a small TP benchmark, we found couple things we can optimize:
1. We call deep_copy so many times when we initialize DTensor.
2. Some shading_prop is not cached successfully.
3. We are still calling redistribute when not necessary.

![image](https://github.com/pytorch/pytorch/assets/6937752/b847d110-eea1-45df-9298-066d0ba07dd7)

![image](https://github.com/pytorch/pytorch/assets/6937752/fc08f564-caed-496b-80d7-275c1dba3806)

![image](https://github.com/pytorch/pytorch/assets/6937752/fdc06cc4-a4ba-48e8-a118-c041bbd04f5e)


So we want to:
1. Remove the deep_copy, this is something risky so it's RFC here.
2. Somehow the op_schema gets changed during sharding_op propogation, so we store a string version of it before passing it to sharding_prop. Ideally we want to figure out why `op_schema` gets changed, but looks like in both index and detach/view op, all get changed, it might take more time to debug.
3. It turns out that Sometimes, DTensor has mem_format to be None (not contiguous) and this will lead to redistribute get triggered.


[ghstack-poisoned]

fduwjj committed:
Update on "[TP][DTensor Perf] Some perf improvement to reduce DTensor CPU overhead"


By inspecting a small TP benchmark, we found couple things we can optimize:
1. We call deep_copy so many times when we initialize DTensor.
2. Some shading_prop is not cached successfully.
3. We are still calling redistribute when not necessary.

![image](https://github.com/pytorch/pytorch/assets/6937752/b847d110-eea1-45df-9298-066d0ba07dd7)

![image](https://github.com/pytorch/pytorch/assets/6937752/fc08f564-caed-496b-80d7-275c1dba3806)

![image](https://github.com/pytorch/pytorch/assets/6937752/fdc06cc4-a4ba-48e8-a118-c041bbd04f5e)


So we want to:
1. Remove the deep_copy, this is something risky so it's RFC here.
2. Somehow the op_schema gets changed during sharding_op propogation, so we store a string version of it before passing it to sharding_prop. Ideally we want to figure out why `op_schema` gets changed, but looks like in both index and detach/view op, all get changed, it might take more time to debug.
3. It turns out that Sometimes, DTensor has mem_format to be None (not contiguous) and this will lead to redistribute get triggered.


[ghstack-poisoned]

fduwjj committed:
Update on "[TP][DTensor Perf] Some perf improvement to reduce DTensor CPU overhead"


By inspecting a small TP benchmark, we found couple things we can optimize:
1. We call deep_copy so many times when we initialize DTensor.
2. Some shading_prop is not cached successfully.
3. We are still calling redistribute when not necessary.

![image](https://github.com/pytorch/pytorch/assets/6937752/b847d110-eea1-45df-9298-066d0ba07dd7)

![image](https://github.com/pytorch/pytorch/assets/6937752/fc08f564-caed-496b-80d7-275c1dba3806)

![image](https://github.com/pytorch/pytorch/assets/6937752/fdc06cc4-a4ba-48e8-a118-c041bbd04f5e)


So we want to:
1. Remove the deep_copy, and we now make placements a tuple so we are sure it's immutable.
2. Somehow the op_schema gets changed during sharding_op propogation, so we store a hash version of it before passing it to sharding_prop. Ideally we want to figure out why `op_schema` gets changed, but looks like in both index and detach/view op, all get changed, it might take more time to debug.
3. Also when we do hashing of op_schema, we want to hash the entire args_schema not just the args_spec which only contains the DTensorSpec from args which are Dtensors.
4. It turns out that sometimes, DTensor has mem_format to be None (not contiguous) and this will lead to redistribute get triggered, so that we only need to compare type/shape and stride in the metadata.

Also we need to ensure _Partial and Shard have different hash value in the DTensorSpec.

![image](https://github.com/pytorch/pytorch/assets/6937752/321e6890-1ab6-4975-adc9-524c6ef9a76b)


[ghstack-poisoned]

wanchaol reviewed (approved):
lgtm,  have only one question

fduwjj commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -351,7 +351,7 @@ def __eq__(self, other: object) -> bool:
 
     def __hash__(self) -> int:
-        return hash(self.reduce_op)
+        return 1 + hash(self.reduce_op)
```
oh, because if we don't have this one, Partial will have the the same hash as shard.

For Partial(SUM) has exactly same hash value as Shard(0). You can see the screenshot I listed in PR summary.

wanchaol replied:
I see, did you test `Shard(1)` and `Partial(SUM)` too and they are still different?

fduwjj commented:
#pytorchbot merge

fduwjj commented:
@pytorchbot merge

fduwjj commented:
@pytorchbot rebase

fduwjj committed:
Update on "[TP][DTensor Perf] Some perf improvement to reduce DTensor CPU overhead"


By inspecting a small TP benchmark, we found couple things we can optimize:
1. We call deep_copy so many times when we initialize DTensor.
2. Some shading_prop is not cached successfully.
3. We are still calling redistribute when not necessary.

![image](https://github.com/pytorch/pytorch/assets/6937752/b847d110-eea1-45df-9298-066d0ba07dd7)

![image](https://github.com/pytorch/pytorch/assets/6937752/fc08f564-caed-496b-80d7-275c1dba3806)

![image](https://github.com/pytorch/pytorch/assets/6937752/fdc06cc4-a4ba-48e8-a118-c041bbd04f5e)


So we want to:
1. Remove the deep_copy, and we now make placements a tuple so we are sure it's immutable.
2. Somehow the op_schema gets changed during sharding_op propogation, so we store a hash version of it before passing it to sharding_prop. Ideally we want to figure out why `op_schema` gets changed, but looks like in both index and detach/view op, all get changed, it might take more time to debug.
3. Also when we do hashing of op_schema, we want to hash the entire args_schema not just the args_spec which only contains the DTensorSpec from args which are Dtensors.
4. It turns out that sometimes, DTensor has mem_format to be None (not contiguous) and this will lead to redistribute get triggered, so that we only need to compare type/shape and stride in the metadata.

Also we need to ensure _Partial and Shard have different hash value in the DTensorSpec.

![image](https://github.com/pytorch/pytorch/assets/6937752/321e6890-1ab6-4975-adc9-524c6ef9a76b)


[ghstack-poisoned]

fduwjj commented:
manually rebase.

fduwjj commented:
@pytorchbot merge

fduwjj commented:
One PR which fixes the broke periodic test caused by this PR is here: https://github.com/pytorch/pytorch/pull/107181/files
